renku/core/management/datasets.py

```diff
@@ -274,141 +278,195 @@ def _add_from_url(self, dataset, path, url, link=False, **kwargs):
-
-            # Submodule in python git does some custom magic that does not
-            # allow for relative URLs, so we call the git function directly
-            self.repo.git.submodule([
-                'add', '--force', '--name', submodule_name, url,
-                submodule_path.relative_to(self.path).as_posix()
-            ])
-
-        src = submodule_path / (target or '')
-
-        if target and relative_to:
-            relative_to = Path(relative_to)
-            if relative_to.is_absolute():
-                assert u.scheme in {
-                    '', 'file'
-                }, 'Only relative paths can be used with URLs.'
-                target = (Path(url).resolve().absolute() / target).relative_to(
-                    relative_to.resolve()
+        dataset_path = self.path / dataset_path
+
+        sources = self._resolve_paths(u.path, sources)
+        destination = destination or Path('.')
+        destination = self._resolve_paths(dataset_path, [destination]).pop()
+
+        dst_root = self.path / dataset_path / destination
+
+        # Get all files from repo that match sources
+        files = set()
+        for file in repo.head.commit.tree.traverse():
+            path = file.path
+            result = self._get_src_and_dst(path, repo_path, sources, dst_root)
+
+            if result:
+                files.add(result)
+
+        # Create metadata and move files to dataset
+        results = []
+        client = LocalClient(repo_path)
+
+        paths = set()
+        for path, src, _ in files:
+            if src.is_dir():
+                continue
+            if src.is_symlink():
+                path = str(src.resolve().relative_to(repo_path))
+            paths.add(path)
+        self._fetch_lfs_files(str(repo_path), paths)
+
+        for path, src, dst in files:
+            if not src.is_dir():
+                creators = []
+                # grab all the creators from the commit history
+                for commit in repo.iter_commits(paths=path):
+                    creator = Creator.from_commit(commit)
+                    if creator not in creators:
+                        creators.append(creator)
+
+                if u.scheme in ('', 'file'):
+                    dst_url = None
+                elif path:
+                    dst_url = '{}/{}'.format(url, path)
+                else:
+                    dst_url = url
+
+                base = DatasetFile.from_revision(
+                    client, path=path, url=url, dataset=None, parent=None
+                )
+
+                results.append({
+                    'path': dst.relative_to(self.path),
+                    'url': dst_url,
+                    'creator': creators,
+                    'dataset': dataset.name,
+                    'parent': self,
+                    'source': base
+                })
+
+                dst.parent.mkdir(parents=True, exist_ok=True)
+                shutil.copy2(str(src), str(dst))
```

Comment:
:+1:  for using copy2 but that might be a problem due to copystat. To explain a bit why - I've seen people using renku on mounted filesystems and since copy2 will try to copy metadata through copystat accessing permission bits can cause this function to fail. In other places, for this reason, we just use copy instead of copy2.